unixcrypt-edited/doc/unixcrypt-breaker.1
.\" Copyright (C) 2006-2013 Peter Selinger.
.\" This file is part of unixcrypt-breaker. It is free software and it is
.\" distributed under the terms of the GNU general public license. See
.\" the file COPYING for details.
.\" $Id: unixcrypt-breaker.1.in 233 2007-10-03 03:33:43Z selinger $
.TH UNIXCRYPT-BREAKER 1 "June 2013" "Version 0.2" "Encryption"
.SH NAME
unixcrypt-breaker \- break unix "crypt" encryption
.SH SYNOPSIS

.nf
.B unixcrypt-breaker [\fIoptions\fP] -s \fIstatfile\fP -c \fIcorpusfile\fP...
.B unixcrypt-breaker [\fIoptions\fP] -s \fIstatfile\fP \fIcipherfile\fP...
.fi
.SH DESCRIPTION

\fBunixcrypt-breaker\fP is an automated tool for breaking the
encryption of the old unix \fBcrypt\fP(1) utility.  It can guess
substantial portions of the plaintext without previous knowledge of
the key.

It is well-known that unix \fBcrypt\fP provides only weak security.
Existing software for breaking it includes the Crypt Breaker's
Workbench (http://axion.physics.ubc.ca/cbw.html). Unlike the Crypt
Breaker's Workbench, \fBunixcrypt-breaker\fP is fully automatic and
does not require user interaction. Rather than relying on the user to
guess likely plaintexts, \fBunixcrypt-breaker\fP uses a statistical
model for guessing plausible plaintexts.

To guess the plaintext correctly, \fBunixcrypt-breaker\fP must first
be primed with a body of sample data (the "corpus") that is presumed
to be similar to the unknown plaintext. For example, if the plaintext
is assumed to be HTML data, then \fBunixcrypt-breaker\fP should be
primed with a corpus of HTML data.  

Note that \fBunixcrypt-breaker\fP can be used to extract a partial
plaintext, but cannot currently be used to extract the actual
encryption key used.
.SH OPERATION

\fBunixcrypt-breaker\fP operates in two distinct steps: \fIsampling\fP
and \fIdecryption\fP. During the sampling step (specified with the
\fB-c\fP option, corresponding to the first form of usage above), data
is read from a corpus of one or more sample plaintext files, and the
resulting statistics are written to a file specified with the \fB-s\fP
option. This file is then used during the decryption step (the second
form of usage above) to partially decrypt one or more ciphertext
files. If more than one file is specified for decryption, they must
have been encrypted with the same key.

It is important to note that \fBunixcrypt-breaker\fP only produces an
approximation to the plaintext. The percentage of bytes correctly
guessed depends on many factors, such as the file type, and the
quality of the corpus of sample data. Performance will be best for
file contents that have high redundancy (such as text).
.SH USAGE EXAMPLE

Suppose file.cpt is a file encrypted with crypt(1), and plain1,
plain2, plain3 are large representative plaintext files of a similar
type. First, sample the plaintext files using the command

.nf
    unixcrypt-breaker -s /tmp/stats.txt -c plain1 plain2 plain3
.fi

This will write some statistical information about the plaintext files
to the file /tmp/stats.txt. 
Then, decrypt file.cpt with the following command: 

.nf
    unixcrypt-breaker -s /tmp/stats.txt file.cpt
.fi

The output is written to file.cpt.guess.
.SH OPTIONS

.TP 15
.B -h, --help
Help. Print usage information and exit.
.TP
.B -V, --version
Print version info and exit.
.TP
.B -L, --license
Print license info and exit.
.TP
.B -c, --corpus
Sampling mode: generate statistics from a corpus of sample
data. The corpus is read from one or more filenames specified on the
command line, or from stdin if no files are specified. The statistics
are written to a file specified with the \fB-s\fP option. For best
performance, the corpus should be as large as possible (several
megabytes would be great), and should be as similar as possible to 
the likely plaintext.
.TP
.B -s \fIfile\fP, --statfile \fIfile\fP
Specify the file to which statistical data is written (in sampling
mode) or from which it is read (in decryption mode). This option is
mandatory.
.TP
.B -o \fIfile\fP, --output \fIfile\fP
Specify a file to write the plaintext guess to. The default is to
append the extension ".guess" to the ciphertext file.
.TP
.B -d \fImode\fP, --display \fImode\fP
Specify the display mode for progress information. Possible choices
are: \fBcolor\fP (default), \fBmono\fP, or \fBnone\fP. The default is
to show a color display that is compatible with color
X-Terminals. \fBmono\fP selects a black-and-white display, and
\fBnone\fP displays no progress information at all. 
.TP
.B -w \fIfile\fP, --wirefile \fIfile\fP
Output optional wiring information to a file. The wirings are internal
to the crypt algorithm and are of interest to experts. 
.TP
.B -m \fIn\fP, --maxdisplay \fIn\fP
Specify how many blocks can fit on the display. This is used when
decrypting multiple files simultanesouly. Regardless of the number of
files given on the command line, at most \fIn\fP blocks will
be displayed simultaneously. The default is 1.
.TP
.B --
End of options. Any remaining arguments are interpreted as filenames.
.PD
.SH VERSION

0.2
.SH AUTHOR

Peter Selinger <selinger at users.sourceforge.net>
.SH COPYRIGHT

Copyright (C) 2006-2013 Peter Selinger

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307,
USA. See also http://www.gnu.org/.


unixcrypt-edited/NEWS
June 20, 2013: Release 0.2

 Added ability to simultaneously decrypt multiple files that were
 encrypted with the same key. Fixed a memory overrun bug on 64-bit
 machines.

February 1, 2008: Release 0.1

 First public release.


unixcrypt-edited/src/unixcrypt-breaker.c
/* Copyright (C) 2006-2013 Peter Selinger.  
   This file is part of unixcrypt-breaker. It is free software and it
   is distributed under the terms of the GNU general public license.
   There is no warranty. See the file COPYING for details. */

/* ---------------------------------------------------------------------- */
/* user interface */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <stdio.h>
#include <getopt.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>

#include "stats.h"
#include "block.h"

#define NAME "unixcrypt-breaker"

/* ---------------------------------------------------------------------- */
/* print usage information */

static void usage(FILE *fout) {
  fprintf(fout, ""NAME" "VERSION". Break unix \"crypt\" encryption.\n\n");
  fprintf(fout, 
  "Usage: "NAME" [options] -s <statfile> -c <corpusfile>...\n"
  "       "NAME" [options] -s <statfile> <cipherfile>...\n"
  "\n"
  "The code-breaking algorithm relies on statistical information from a\n"
  "corpus of sample data that is presumed similar to the unknown\n"
  "plaintext. To use this program, you must first run "NAME"\n"
  "with the -c option on a corpus of sample data to create a statfile\n"
  "(see the first form of usage above). You may then use the resulting\n"
  "statfile to attempt decryption of the cipherfile (second form).\n"
  "\n"
  "Options:\n"
  " -h, --help            print this help message and exit\n"
  " -V, --version         print version info and exit\n"
  " -L, --license         print license info and exit\n"
  " -D, --debug           raise debugging level\n"
  " -c, --corpus          create statfile from one or more files of sample\n"
  "                       data, or from stdin if no filenames are given.\n"
  " -s, --statfile <file> specify path to statfile (mandatory)\n"
  " -d, --display <mode>  progress display: color (default), mono, scroll, or none\n"
  " -m, --maxdisplay <n>  how many blocks fit on the display (default 1)\n"
  " -o, --output <file>   select output file (default: <cipherfile>.guess)\n"
  " -w, --wirefile <file> output wiring information (advanced) to file\n"
  " --                    end of options, filenames follow\n");
}

/* print version and copyright information */
static void version(FILE *fout) {
  fprintf(fout, ""NAME" "VERSION". Break unix \"crypt\" encryption.\n");
  fprintf(fout, "Copyright (C) 2006-2013 Peter Selinger.\n");
}

static void license(FILE *fout) {
  fprintf(fout, ""NAME" "VERSION". Break unix \"crypt\" encryption.\n");
  fprintf(fout, "Copyright (C) 2006-2013 Peter Selinger.\n");
  fprintf(fout, "\n");
  fprintf(fout,
  "This program is free software; you can redistribute it and/or modify\n"
  "it under the terms of the GNU General Public License as published by\n"
  "the Free Software Foundation; either version 2 of the License, or\n"
  "(at your option) any later version.\n"
  "\n"
  "This program is distributed in the hope that it will be useful,\n"
  "but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
  "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
  "GNU General Public License for more details.\n"
  "\n"
  "You should have received a copy of the GNU General Public License\n"
  "along with this program; if not, see <http://www.gnu.org/licenses/>.\n"
  );
}

static struct option longopts[] = {
  {"help",         0, 0, 'h'},
  {"version",      0, 0, 'V'},
  {"license",      0, 0, 'L'},
  {"corpus",       0, 0, 'c'},
  {"statfile",     1, 0, 's'},
  {"output",       1, 0, 'o'},
  {"display",      1, 0, 'd'},
  {"maxdisplay",   1, 0, 'm'},
  {"debug",        0, 0, 'D'},
  {"wirefile",     1, 0, 'w'},
  {0, 0, 0, 0}
};

static char *shortopts = "-hVLcs:o:d:Dw:m:";

static int digest_corpus(char *statfile, char **av, int ac) {
  FILE *fstat;
  FILE *fin;
  int i;
  stats_t *t;
  
  fstat = fopen(statfile, "w");
  if (!fstat) {
    fprintf(stderr, ""NAME": %s: %s\n", statfile, strerror(errno));
    exit(1);
  }      

  t = stats_new();
  if (!t) {
    fprintf(stderr, ""NAME": %s\n", strerror(errno));
    exit(1);
  }
  
  if (ac == 0) {  /* read from stdin */
    stats_digest(stdin, t);
  } else {
    for (i=0; i<ac; i++) {
      fin = fopen(av[i], "r");
      if (!fin) {
	fprintf(stderr, ""NAME": %s: %s, skipping\n", av[i], strerror(errno));
      } else {
	stats_digest(fin, t);
	fclose(fin);
      }
    }
  }
  stats_write(fstat, t);
  stats_free(t);
  fclose(fstat);
  return 0;
}

/* Read N streams of ciphertext from fcipher_x[], and write N streams
   of plaintext guesses to fout_x[]. Display progress info according
   to display_mode and debug. fwire, if non-NULL, is a stream to write
   wiring info to. fr_x[] is an array of statistics. Display_max
   specifies how many blocks can be simultaneously displayed on the
   screen. */
static void unixcrypt_breaker_streams(int N, FILE **fcipher_x, FILE *fwire, FILE **fout_x, stats_t **fr_x, int display_mode, int debug, int display_max) {
  int b, j;
  int block_cipher_x[N][256];
  int block_plain_x[N][256];
  int wires_known[256];
  int n_x[N];
  int file_ended_x[N];  /* has file j ended? */
  int nopen;            /* number of non-ended files */

  nopen = N;
  memset(file_ended_x, 0, sizeof(file_ended_x));

  b=0;
  while (1) {
    for (j=0; j<N; j++) {
      if (!file_ended_x[j]) {
	n_x[j] = read_block(fcipher_x[j], block_cipher_x[j]);
      } else {
	n_x[j] = 0;
      }
    }
    for (j=0; j<N; j++) {
      unmask_block(n_x[j], block_cipher_x[j]);
      erase_block(block_plain_x[j]);
    }
    process_block(N, fr_x, n_x, block_cipher_x, block_plain_x, wires_known, display_mode, debug, b, display_max);
    if (display_mode == DISPLAY_SCROLL) {
      for (j=0; j<N; j++) {
	if (n_x[j] != 0) {
	  fprintf(stdout, "File %d, block %d:\n", j, b);
	  write_block_hex(stdout, n_x[j], block_plain_x[j], NULL, 0);
	  fprintf(stdout, "\n");
	}
      }
    }
    for (j=0; j<N; j++) {
      if (n_x[j] != 0) {
	write_block(fout_x[j], n_x[j], block_plain_x[j]);
	fflush(fout_x[j]);
      }
    }
    if (fwire) {
      write_wires_long(fwire, wires_known, b);
      fflush(fwire);
    }
    for (j=0; j<N; j++) {
      if (!file_ended_x[j] && n_x[j] < 256) {
	file_ended_x[j] = 1;
	nopen--;
      }
    }
    if (nopen == 0) {
      break;
    }
    b++;
  }
}

/* Decipher N files from cipherfile_x[] to outfile_x[], using
   statistics files statfile_x[]. All files must be encrypted with the
   same key. wirefile is optional; if present, wiring information will
   be written to it. Print progress and debug info according to
   display_mode and debug. Display_max specifies how many blocks can
   be simultaneously displayed on the screen. */
static int unixcrypt_breaker(int N, char **cipherfile_x, char **statfile_x, char *wirefile, char **outfile_x, int display_mode, int debug, int display_max) {
  int j;

  FILE *fcipher_x[N];
  FILE *fwire;
  FILE *fstat;
  FILE *fout_x[N];
  stats_t *fr_x[N];

  /* open and read statfiles. */
  for (j=0; j<N; j++) {
    /* optimize in case the same statfile is repeated */
    if (j>0 && strcmp(statfile_x[j], statfile_x[j-1]) == 0) {
      fr_x[j] = fr_x[j-1];
      continue;
    }
    fstat = fopen(statfile_x[j], "r");
    if (!fstat) {
      fprintf(stderr, ""NAME": %s: %s\n", statfile_x[j], strerror(errno));
      exit(1);
    }      
    fr_x[j] = stats_read(fstat);
    if (!fr_x[j]) {
      fprintf(stderr, ""NAME": %s\n", strerror(errno));
      return 1;
    } else if (fr_x[j] == STATS_EFORMAT) {
      fprintf(stderr, ""NAME": %s: bad file format\n", statfile_x[j]);
      return 1;
    }
    fclose(fstat);
  }

  /* open cipherfiles */
  for (j=0; j<N; j++) {
    fcipher_x[j] = fopen(cipherfile_x[j], "r");
    if (!fcipher_x[j]) {
      fprintf(stderr, ""NAME": %s: %s\n", cipherfile_x[j], strerror(errno));
      exit(1);
    }
  }
  
  /* open outfiles */
  for (j=0; j<N; j++) {
    fout_x[j] = fopen(outfile_x[j], "w");
    if (!fout_x[j]) {
      fprintf(stderr, ""NAME": %s: %s\n", outfile_x[j], strerror(errno));
      exit(1);
    }
  }

  /* open wirefile, if any */
  if (wirefile) {
    fwire = fopen(wirefile, "w");
    if (!fwire) {
      fprintf(stderr, ""NAME": %s: %s\n", wirefile, strerror(errno));
      exit(1);
    }
  } else {
    fwire = NULL;
  }

  unixcrypt_breaker_streams(N, fcipher_x, fwire, fout_x, fr_x, display_mode, debug, display_max);

  for (j=0; j<N; j++) {
    fclose(fcipher_x[j]);
    fclose(fout_x[j]);
  }
  if (fwire) {
    fclose(fwire);
  }
  return 0;
}

int main(int ac, char *av[]) {
  int c, j;
  const char *guessext = ".guess";
  char *statfile_y = NULL;
  char *outfile_y = NULL;

  /* command line options and defaults */
  int display_mode = DISPLAY_COLOR;  
  int corpus_mode = 0;   /* 1=on, 0=off */
  char *wirefile = NULL;
  char **outfile_x = NULL;
  char **cipherfile_x = NULL;
  char **statfile_x = NULL;
  int N = 0; /* current size of outfile_x, cipherfile_x, statfile_x arrays */
  int debug = 0;
  int display_max = 1;

  while ((c = getopt_long(ac, av, shortopts, longopts, NULL)) != -1) {
    switch (c) {
    case 'h':
      usage(stdout);
      exit(0);
      break;
    case 'V':
      version(stdout);
      exit(0);
      break;
    case 'L':
      license(stdout);
      exit(0);
      break;
    case 'c':
      corpus_mode = 1;
      break;
    case 1:
      N++;
      cipherfile_x = realloc(cipherfile_x, N * sizeof(*cipherfile_x));
      if (!cipherfile_x) {
	goto error;
      }
      statfile_x = realloc(statfile_x, N * sizeof(*statfile_x));
      if (!statfile_x) {
	goto error;
      }
      outfile_x = realloc(outfile_x, N * sizeof(*outfile_x));
      if (!outfile_x) {
	goto error;
      }
      statfile_x[N-1] = statfile_y;
      cipherfile_x[N-1] = optarg;
      outfile_x[N-1] = outfile_y;
      outfile_y = NULL;
      break;
    case 's':
      statfile_y = optarg;
      for (j=0; j<N; j++) {
	if (statfile_x[j] == NULL) {
	  statfile_x[j] = statfile_y;
	} else {
	  break;
	}
      }
      break;
    case 'o':
      outfile_y = optarg;
      break;
    case 'd':
      if (strcmp(optarg, "color") == 0) {
	display_mode = DISPLAY_COLOR;
      } else if (strcmp(optarg, "mono") == 0) {
	display_mode = DISPLAY_MONO;
      } else if (strcmp(optarg, "scroll") == 0) {
	display_mode = DISPLAY_SCROLL;
      } else if (strcmp(optarg, "none") == 0) {
	display_mode = DISPLAY_NONE;
      } else {
	fprintf(stderr, ""NAME": invalid display mode: %s. Try --help for more info.\n", optarg);
	exit(1);
      }
      break;
    case 'm':
      display_max = atoi(optarg);
      if (display_max < 1) {
	fprintf(stderr, ""NAME": argument to -m must be 1 or more\n");
	exit(1);
      }
      break;
    case 'D':
      debug++;
      break;
    case 'w':
      wirefile = optarg;
      break;
    case '?': default:
      fprintf(stderr, "Try --help for more information.\n");
      exit(1);
      break;
    }
  }

  if (N == 0) {
    fprintf(stderr, ""NAME": no filenames specified. Try --help for more info.\n");
    exit(1);
  }

  /* check for statfile */
  if (statfile_y == NULL) {
    fprintf(stderr, ""NAME": the -s option is mandatory. Try --help for more info.\n");
    exit(1);
  }

  if (corpus_mode) {
    return digest_corpus(statfile_y, cipherfile_x, N);
  }

  /* check for unused outfile */
  if (outfile_y && N > 1) {
    fprintf(stderr, ""NAME": -o option at the end of file list not allowed\n");
    exit(1);
  } else if (outfile_y && N == 1) {  /* for backward compatibility */
    outfile_x[0] = outfile_y;
  }

  for (j=0; j<N; j++) {
    if (outfile_x[j] == NULL) {
      outfile_x[j] = malloc(strlen(cipherfile_x[j])+strlen(guessext)+1);
      if (!outfile_x[j]) {
	goto error;
      }
      strcpy(outfile_x[j], cipherfile_x[j]);
      strcat(outfile_x[j], guessext);
    }
  }

  return unixcrypt_breaker(N, cipherfile_x, statfile_x, wirefile, outfile_x, display_mode, debug, display_max);

 error:
  fprintf(stderr, ""NAME": %s\n", strerror(errno));
  exit(1);
}
